src/tools/wasm-fuzz-lattices.cpp

```diff
@@ -35,14 +42,421 @@ using namespace analysis;
 // Helps printing error messages.
 std::string LatticeComparisonNames[4] = {
   "No Relation", "Equal", "Less", "Greater"};
-std::string LatticeComparisonSymbols[4] = {"?", "=", "<", ">"};
 
 uint64_t getSeed() {
   // Return a (truly) random 64-bit value.
   std::random_device rand;
   return std::uniform_int_distribution<uint64_t>{}(rand);
 }
 
+// Actually a pointer to `L::ElementImpl`, but we erase the type to avoid
+// getting into a situation where `L` satisfying `Lattice` or `FullLattice`
+// circularly requires that `L` satisfies `Lattice` or `FullLattice`. C++ does
+// not allow concepts to depend on themselves. Also make the pointer copyable to
+// satisfy that constraint on lattice elements.
+template<typename L>
+struct RandomElement : std::unique_ptr<void, void (*)(void*)> {
+  RandomElement() = default;
+
+  RandomElement(typename L::ElementImpl&& other)
+    : std::unique_ptr<void, void (*)(void*)>(
+        new typename L::ElementImpl(std::move(other)),
+        [](void* e) { delete static_cast<typename L::ElementImpl*>(e); }) {}
+
+  RandomElement(const RandomElement& other)
+    : RandomElement([&]() {
+        auto copy = *other;
+        return copy;
+      }()) {}
+
+  RandomElement(RandomElement&& other) = default;
+  RandomElement& operator=(const RandomElement& other) {
+    if (this != &other) {
+      new (this) RandomElement(other);
+    }
+    return *this;
+  }
+
+  RandomElement& operator=(RandomElement&& other) = default;
+
+  typename L::ElementImpl& operator*() {
+    return *static_cast<typename L::ElementImpl*>(get());
+  }
+
+  const typename L::ElementImpl& operator*() const {
+    return *static_cast<const typename L::ElementImpl*>(get());
+  }
+
+  typename L::ElementImpl* operator->() { return &*(*this); }
+
+  const typename L::ElementImpl* operator->() const { return &*(*this); }
+};
+
+struct RandomFullLattice {
+  // The inner lattice and lattice element types. These must be defined later
+  // because they depend on `RandomFullLattice` satisfying `FullLattice`, but
+  // that requires the type to be complete.
+  struct L;
+  struct ElementImpl;
```

Comment:
Wait, C++ lets you do a forward declaration in the struct definition, and you can define the forward-declared thing later outside? So you could even define it in another file..? How can that work?

Reply:
Yep, just like for a method. The type isn't complete until you define it, of course, which is why many of the methods below also have to be defined out of line, and after `L` and `ElementImpl` have been defined.

Reply:
Unfortunately you can't forward-declare a `using` statement, which is why these have to be expressed with structs and trivial inheritance.